interpreters/wamr/wamr_custom_init.c

```diff
@@ -0,0 +1,657 @@
+              case 'c':
+                {
+                  if (long_ctr < 2)
+                    {
+                      pos += sizeof(int32_t);
+                    }
+                  else
+                    {
+                      pos += sizeof(int64_t);
+                    }
+                  break;
+                }
+
+              case 'e':
+              case 'E':
+              case 'g':
+              case 'G':
+              case 'f':
+              case 'F':
+                {
+                  pos += sizeof(double);
+                  break;
+                }
+
+              case 's':
+                {
+                  if (to_native)
+                    {
+                      *(uintptr_t *)pos =
+                          (uintptr_t)addr_app_to_native(*(uintptr_t *)pos);
+                    }
+                  else
+                    {
+                      *(uintptr_t *)pos =
+                          (uintptr_t)addr_native_to_app(*(uintptr_t *)pos);
+                    }
+                  pos += sizeof(uintptr_t);
+                  break;
+                }
+
+              default:
+                  break;
+          }
+
+        might = 0;
+      }
+
+  still_might:
+      ++format;
+  }
+}
+
+
+#define va_list_string2native(exec_env, format, ap) \
+  va_list_string2conv(exec_env, format, ap, true)
+
+#define va_list_string2app(exec_env, format, ap) \
+  va_list_string2conv(exec_env, format, ap, false)
+
+static void
+scanf_begin(wasm_module_inst_t module_inst, va_list ap)
+{
+  uintptr_t *apv = *(uintptr_t **)&ap;
+  if (apv == NULL)
+    {
+      return;
+    }
+  while (*apv != 0)
+    {
+      *apv = (uintptr_t)addr_app_to_native(*apv);
+      apv++;
+    }
+}
+
+static void
+scanf_end(wasm_module_inst_t module_inst, va_list ap)
+{
+  uintptr_t *apv = *(uintptr_t **)&ap;
+  if (apv == NULL)
+    {
+      return;
+    }
+  while (*apv != 0)
+    {
+      *apv = (uintptr_t)addr_native_to_app((void *)*apv);
+      apv++;
+    }
+}
+
+static pthread_mutex_t g_compare_mutex = PTHREAD_MUTEX_INITIALIZER;
+static wasm_exec_env_t g_compare_env;
+static void           *g_compare_func;
+
+static int
+compare_proxy(const void *a, const void *b)
+{
+  wasm_module_inst_t module_inst = get_module_inst(g_compare_env);
+  uint32_t argv[2];
+
+  argv[0] = addr_native_to_app((void *)a);
+  argv[1] = addr_native_to_app((void *)b);
+
+  return wasm_runtime_call_indirect(g_compare_env,
+           (uint32_t)addr_native_to_app(g_compare_func), 2, argv) ?
+             argv[0] : 0;
+}
+
+#ifndef GLUE_FUNCTION_qsort
+#define GLUE_FUNCTION_qsort
+void glue_qsort(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2,
+                uintptr_t parm3, uintptr_t parm4)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  pthread_mutex_lock(&g_compare_mutex);
+  g_compare_env = env;
+  g_compare_func = parm4;
+  qsort((FAR void *)parm1, (size_t)parm2,
+        (size_t)parm3, compare_proxy);
```

Comment:
this seems broken as the user callback might not return. or it might even call qsort.